### PR TITLE
fix: honor -rl fallback and preserve -rls duration units

### PR DESCRIPTION
**Pull request by wh0amibjm**
## Proposed changes
/claim #1434

This PR fixes two concrete issues behind `-rl` / `-rls` not behaving as expected:

1. **Global rate limit fallback was not consistently applied**
   - `buildMultiRateLimiter` now starts from the global `-rl` value for each source.
   - If a source has a custom `-rls` entry, it overrides global behavior.
   - Passing nil custom limits no longer risks a nil dereference.

2. **Per-source duration units from `-rls` were being dropped**
   - Introduced `subscraping.RateLimitSpec{MaxCount, Duration}`.
   - Runner now preserves both `MaxCount` and parsed `Duration` (e.g. `2/m`, `10/s`) when building custom limits.
   - Passive limiter now uses the configured per-source duration instead of forcing `time.Second`.

The fix is intentionally minimal and scoped to rate-limit plumbing.

### Proof
Focused tests added to validate behavior and prevent regressions:

- `pkg/passive/passive_ratelimit_test.go`
  - global fallback when source is missing from custom map
  - global fallback when custom map is nil
  - per-source custom duration is respected (timing assertion)
  - unlimited behavior unchanged when no limits are provided
- `pkg/runner/runner_ratelimit_test.go`
  - duration is preserved from parsed `RateLimitMap`
  - zero-duration entries default to 1 second
  - invalid zero-count entries are ignored

Executed:

```bash
go test ./pkg/passive ./pkg/runner -run 'TestBuildMultiRateLimiter|TestSourceRateLimitOrDefault|TestBuildCustomRateLimit' -count=1 -v
```

Output:

```text
=== RUN   TestBuildMultiRateLimiter_AppliesGlobalRateLimitWhenCustomMissing
--- PASS: TestBuildMultiRateLimiter_AppliesGlobalRateLimitWhenCustomMissing (0.00s)
=== RUN   TestBuildMultiRateLimiter_AppliesGlobalRateLimitWhenCustomIsNil
--- PASS: TestBuildMultiRateLimiter_AppliesGlobalRateLimitWhenCustomIsNil (0.00s)
=== RUN   TestBuildMultiRateLimiter_UsesCustomDuration
--- PASS: TestBuildMultiRateLimiter_UsesCustomDuration (0.15s)
=== RUN   TestBuildMultiRateLimiter_UnlimitedWhenNoRateLimitProvided
--- PASS: TestBuildMultiRateLimiter_UnlimitedWhenNoRateLimitProvided (0.00s)
=== RUN   TestSourceRateLimitOrDefault
--- PASS: TestSourceRateLimitOrDefault (0.00s)
PASS
ok   github.com/projectdiscovery/subfinder/v2/pkg/passive 0.252s
=== RUN   TestBuildCustomRateLimit
--- PASS: TestBuildCustomRateLimit (0.00s)
PASS
ok   github.com/projectdiscovery/subfinder/v2/pkg/runner 0.084s
```

## Checklist
- [x] Pull request is created against the `dev` branch
- [x] All checks passed locally for affected areas
- [x] Tests added to prove the fix is effective
- [ ] Documentation added (not required for this internal bug fix)

## Summary by CodeRabbit

* **New Features**
  * Configurable per-source rate limit durations so each data source can have independent timing in addition to global limits.

* **Bug Fixes**
  * Protects against negative or zero rate values and correctly falls back to unlimited or default behaviors.
  * Ensures per-source durations are respected rather than always using a fixed interval.

* **Refactor**
  * Centralized rate-limit initialization for more consistent behavior.

* **Tests**
  * Added unit tests validating multi-source rate-limit behaviors and edge cases.

### CODE REVIEW

File: pkg/runner/runner.go [hunk 1]
```diff
@@ -57,19 +59,32 @@ func NewRunner(options *Options) (*Runner, error) {
...
+func buildCustomRateLimit(rateLimits map[string]goflags.RateLimit) *subscraping.CustomRateLimit {
+	customRateLimit := &subscraping.CustomRateLimit{
+		Custom: mapsutil.SyncLockMap[string, subscraping.RateLimitSpec]{
+			Map: make(map[string]subscraping.RateLimitSpec),
 		},
 	}
 
-	for source, sourceRateLimit := range options.RateLimits.AsMap() {
+	for source, sourceRateLimit := range rateLimits {
 		if sourceRateLimit.MaxCount > 0 && sourceRateLimit.MaxCount <= math.MaxUint {
```
[the remark this answers is not in the document]
Comment (wh0amibjm) [reply pkg/runner/runner.go]: Addressed in `a313a7f`:\n- removed tautological `math.MaxUint` check\n- normalized custom source keys with `strings.ToLower(source)`\n- switched SyncLockMap initializations to `mapsutil.NewSyncLockMap(..., mapsutil.WithMap(...))` in tests and constructor path.

**Comment by wh0amibjm** [comment]
Quick follow-up on merge readiness:\n\n- CodeRabbit + Neo are green\n- fixes for  fallback and  duration-unit preservation are in\n- targeted tests were run and passed in prior update\n\nCould a maintainer take a final review pass for merge when convenient? 🙏

**Comment by wh0amibjm** [comment]
Correction to previous note: the fixes are for global rate-limit fallback and per-source duration-unit preservation. Everything else in the status update remains the same.

**Comment by dogancanbakir** [comment]
Hi, thanks for your interest in contributing! Just a heads up, we ask contributors to work on **1 active issue at a time** ([see](https://github.com/projectdiscovery/oss-bounty-program?tab=readme-ov-file#how-to-participatee)).

Also, we welcome AI-assisted development, but submissions must be complete, tested, and ready to merge. Please also make sure to fill out the PR template with proof that your changes work.

We're closing this PR along with your other open submissions. Once you're ready, feel free to pick one issue to focus on and resubmit; we'd be happy to review it.

Appreciate your understanding!
